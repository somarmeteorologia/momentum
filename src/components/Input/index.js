import React, { memo, useState, useContext } from 'react'
import styled, { css, ThemeContext } from 'styled-components'
import PropTypes from 'prop-types'
import { ifProp, switchProp, prop } from 'styled-tools'

const iconAlign = {
  left: 'left',
  right: 'right'
}

const typing = {
  text: 'text',
  password: 'password',
  email: 'email'
}

const Container = styled.label`
  display: flex;
  flex-direction: column;
  position: relative;
  width: ${ifProp('full', '100%', '240px')};
`

const Label = styled.span`
  margin-bottom: 8px;
  font-family: ${prop('theme.font.family.inter')};
  font-size: ${prop('theme.font.size.twelve')};
  font-weight: ${prop('theme.font.weight.bold')};
  color: ${ifProp(
  'disabled',
  prop('theme.field.text.disabled'),
  prop('theme.field.text.primary')
)};

  ${ifProp(
  'required',
  css`
      &::after {
        content: '*';
        color: ${ifProp(
    'disabled',
    prop('theme.field.text.disabled'),
    prop('theme.field.text.danger')
  )};
      }
    `
)}
`

const Inputable = styled.input`
  height: 40px;
  padding: 11px 14px;
  font-family: ${prop('theme.font.family.inter')};
  font-size: ${prop('theme.font.size.fourteen')};
  color: ${prop('theme.field.text.primary')};
  background-color: ${prop('theme.field.bg.primary')};
  border-radius: ${prop('theme.border.radius.four')};
  border: ${ifProp(
  'hasError',
  prop('theme.field.border.danger'),
  prop('theme.field.border.primary')
)};
  cursor: ${ifProp('disabled', 'not-allowed', 'initial')};

  ${ifProp(
  'icon',
  switchProp('iconAlign', {
    [iconAlign.left]: css`
        padding-left: 36px;
      `,
    [iconAlign.right]: css`
        padding-right: 36px;
      `
  }),
  null
)}

  &:focus {
    outline: none;
    border: ${ifProp(
  'hasError',
  prop('theme.field.border.danger'),
  prop('theme.field.border.secondary')
)};
  }

  &::placeholder {
    color: ${ifProp(
  'hasError',
  prop('theme.field.text.danger'),
  prop('theme.field.text.secondary')
)};

    ${ifProp(
  'disabled',
  css`
        color: ${prop('theme.field.text.disabled')};
      `
)};
  }
`

const Message = styled.span`
  margin-top: 8px;
  font-family: ${prop('theme.font.family.inter')};
  font-size: ${prop('theme.font.size.ten')};
  color: ${prop('theme.field.text.danger')};
`

const Content = styled.div`
  display: flex;
  align-items: center;
  position: absolute;
  height: 40px;
  line-height: 40px;

  ${switchProp('iconAlign', {
  left: css`
      left: 16px;
    `,
  right: css`
      right: 16px;
    `
})};
`

export const Input = memo(
  ({
    value,
    label,
    typing,
    onChange,
    placeholder,
    icon,
    iconAlign,
    error,
    success,
    required,
    disabled,
    full,
    ...props
  }) => {
    const { field } = useContext(ThemeContext)
    const [text, setText] = useState(value)

    const whenChange = ({ target }) => {
      const { value } = target

      setText(value)
      onChange(value)
    }

    return (
      <Container full={full}>
        {label && (
          <Label required={required} disabled={disabled}>
            {label}
          </Label>
        )}

        {icon && (
          <Content iconAlign={iconAlign}>
            {icon({
              width: 14,
              height: 14,
              color:
                success && success.has ? field.icon.success : field.icon.primary
            })}
          </Content>
        )}

        <Inputable
          type={typing}
          value={text}
          hasError={error.has}
          icon={icon}
          iconAlign={iconAlign}
          onChange={whenChange}
          placeholder={placeholder}
          required={required}
          disabled={disabled}
          {...props}
        />

        {error.has && <Message>{error.message}</Message>}
      </Container>
    )
  }
)

Input.defaultProps = {
  value: '',
  label: '',
  placeholder: '',
  iconAlign: iconAlign.left,
  typing: typing.text,
  onChange: () => { },
  error: {
    has: false
  },
  success: {
    has: false
  },
  disabled: false,
  required: false,
  full: false
}

Input.propTypes = {
  value: PropTypes.string,
  label: PropTypes.string,
  placeholder: PropTypes.string,
  typing: PropTypes.oneOf([typing.password, typing.text, typing.email]),
  icon: PropTypes.func,
  iconAlign: PropTypes.oneOf([iconAlign.left, iconAlign.right]),
  error: PropTypes.shape({
    has: PropTypes.bool,
    message: PropTypes.string
  }),
  success: PropTypes.shape({
    has: PropTypes.bool,
    message: PropTypes.string
  }),
  required: PropTypes.bool,
  disabled: PropTypes.bool,
  full: PropTypes.bool
}

Input.typing = typing
Input.iconAlign = iconAlign
