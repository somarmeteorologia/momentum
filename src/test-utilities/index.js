import React from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'

import { Theme } from '@components/Theme'

export const renderWithTheme = (ui, theme = Theme.light) => {
  return {
    ...render(<ThemeProvider theme={theme}>{ui}</ThemeProvider>)
  }
}
