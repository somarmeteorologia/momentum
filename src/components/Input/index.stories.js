import React from 'react'
import styled from 'styled-components'
import { storiesOf } from '@storybook/react'

import { GROUPS } from '@helpers/hierarchySeparators'
import { Container } from '@helpers/components/Container'

import { Reset } from '@components/Reset'
import { Input } from '@components/Input'
import { Icon } from '@components/Icon'

const Containerable = styled(Container)`
  flex-wrap: wrap;
`

const Wrapper = styled.div`
  display: flex;
  align-items: flex-end;
`

storiesOf(`${GROUPS.COMPONENTS}|Input`, module)
  .addDecorator(story => (
    <Containerable>
      <Reset />
      {story()}
    </Containerable>
  ))
  .add('Default', () => (
    <>
      <Input placeholder="Your document" />
    </>
  ))
  .add('With label', () => (
    <>
      <Input placeholder="Your document" label="Field label" />
      <Input placeholder="Your document" label="Field label" required />
    </>
  ))
  .add('With error', () => (
    <>
      <Input
        placeholder="Your document"
        error={{ has: true, message: 'Your document' }}
      />
    </>
  ))
  .add('With success', () => (
    <>
      <Input
        placeholder="Your e-mail"
        icon={({ color, width, height }) => (
          <Icon
            name={Icon.name.checked}
            color={color}
            width={width}
            height={height}
          />
        )}
        iconAlign={Input.iconAlign.right}
        success={{ has: true, message: 'Your document' }}
      />
    </>
  ))
  .add('With icon', () => (
    <>
      <Input
        placeholder="Your document"
        icon={({ color, width, height }) => (
          <Icon
            name={Icon.name.checked}
            color={color}
            width={width}
            height={height}
          />
        )}
        iconAlign={Input.iconAlign.left}
      />

      <Input
        placeholder="Your e-mail"
        icon={({ color, width, height }) => (
          <Icon
            name={Icon.name.checked}
            color={color}
            width={width}
            height={height}
          />
        )}
        iconAlign={Input.iconAlign.right}
      />
    </>
  ))
  .add('Types', () => (
    <>
      <Input
        placeholder="Your document"
        icon={({ color, width, height }) => (
          <Icon
            name={Icon.name.checked}
            color={color}
            width={width}
            height={height}
          />
        )}
        iconAlign={Input.iconAlign.left}
      />

      <Input
        placeholder="Your email"
        type={Input.typing.email}
        icon={({ color, width, height }) => (
          <Icon
            name={Icon.name.checked}
            color={color}
            width={width}
            height={height}
          />
        )}
        iconAlign={Input.iconAlign.right}
      />
      <Input
        placeholder="Your password"
        type={Input.typing.password}
        icon={({ color, width, height }) => (
          <Icon
            name={Icon.name.checked}
            color={color}
            width={width}
            height={height}
          />
        )}
        iconAlign={Input.iconAlign.right}
      />
    </>
  ))
  .add('Full', () => (
    <>
      <Input placeholder="Your document" full />
    </>
  ))
  .add('Disabled', () => (
    <Wrapper>
      <Input placeholder="Your document" disabled />
      <Input placeholder="Your document" label="Field label" disabled />
      <Input
        placeholder="Your document"
        label="Field label"
        disabled
        required
      />
    </Wrapper>
  ))
