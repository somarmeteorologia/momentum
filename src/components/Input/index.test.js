import React from 'react'
import { fireEvent } from '@testing-library/react'

import { renderWithTheme } from '../../test-utilities'

import { Input } from '@components/Input'

test('Not expect visual regression', () => {
  const { container } = renderWithTheme(<Input />)

  expect(container.firstChild).toMatchSnapshot()
})

test('Should return only the value when raw is false', () => {
  let onChangeOutput

  const { getByTestId } = renderWithTheme(
    <Input data-testid="input" onChange={value => (onChangeOutput = value)} />
  )

  fireEvent.input(getByTestId('input'), { target: { value: 'John Doe' } })

  expect(onChangeOutput).toBe('John Doe')
})

test('Should return the event object when raw is true', () => {
  let onChangeOutput

  const { getByTestId } = renderWithTheme(
    <Input
      data-testid="input"
      raw={true}
      onChange={({ target }) => (onChangeOutput = target.value)}
    />
  )

  fireEvent.input(getByTestId('input'), { target: { value: 'John Doe' } })

  expect(onChangeOutput).toBe('John Doe')
})
