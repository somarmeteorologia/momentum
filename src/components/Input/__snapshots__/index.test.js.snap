// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Not expect visual regression 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  width: 240px;
}

.c1 {
  height: 40px;
  padding: 11px 14px;
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  color: #19244d;
  background-color: #f1f3fA;
  border-radius: 0.25rem;
  border: 1px solid #b1bbe5;
  cursor: initial;
}

.c1:focus {
  outline: none;
  border: 1px solid #6479cb;
}

.c1::-webkit-input-placeholder {
  color: #6479cb;
}

.c1::-moz-placeholder {
  color: #6479cb;
}

.c1:-ms-input-placeholder {
  color: #6479cb;
}

.c1::placeholder {
  color: #6479cb;
}

<label
  class="c0"
>
  
  <input
    class="c1"
    placeholder=""
    type="text"
    value=""
  />
</label>
`;
